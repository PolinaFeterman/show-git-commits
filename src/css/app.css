.app {
  text-align: center;
}
.header {
  height: 100px;
  background-color: #edf7fc;
  margin-bottom: 20px;
}
