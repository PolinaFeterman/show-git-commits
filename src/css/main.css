@import "bootstrap";

#container {
  margin: 1rem;
}

.repo-title {
  margin-bottom: 1.5rem;
}

.listing {
  margin-bottom: 1rem;
}

.mainDiv{
    margin-top: "1rem"
}
.paginationWrapper{
    width: 30%;
    margin: 0 auto;
}
.hidden{
    display:none;
}
.backWrapper{
    text-align: left;
    

}
.settingsButton{
    margin: 20px;
    border-radius: 10px;
    border: 2px solid #0d6efd;
    width: 200px;
    background-color: #EDF7FC;

}