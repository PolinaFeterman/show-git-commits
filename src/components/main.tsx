import React, { useContext, useEffect, useState } from "react";
import "../css/main.css";
//import generatedGitInfo from '../generatedGitInfo.json';
import { CommitCard } from "./commit-card";
import ReactPaginate from "react-paginate";
import ReactDOM from "react-dom";
import { useParams } from "react-router";
import { fetchCommitsData, PaginationService } from "../services/pagination";
import { Commits } from "./commits";
import { Pagination } from "./pagination";
import { Link, useNavigate } from "react-router-dom";
import { AppCtx } from "../app";
import AppContextInterface from "../interfaces/app-context-interface";
import CommitCardInterface from "../interfaces/commit-card-interface";

export type mainProps = {
  orgName?: string;
  repoName?: string;
  perPage?: number;
};

export const Main = (props: mainProps) => {
  const [repositories, setRepositories] = React.useState([
    { commitDate: "", commitAuthor: "", commitMessage: "", commitUrl: "" },
  ]);
  const [pageOffset, setPageOffset] = React.useState(0);
  const [pageCount, setPageCount] = React.useState(1);
  const [commitsInfoLength, setCommitsInfoLength] = React.useState(0);
  // const [apiError, setApiError] = useState('');
  const { org, repo } = useParams();
  const navigate = useNavigate();
  const appContext = React.useContext(AppCtx) as AppContextInterface;

  useEffect(() => {
    (async () => {
      
      const fetchCommitsData: fetchCommitsData = {
        perPage: appContext.perPage.toString() || "10",
        pageOffset: (pageOffset + 1).toString(),
        orgName: org || "",
        repoName: repo || "",
      };
      const response = await PaginationService.fetchCommits(fetchCommitsData);
      //const responseJson = await response.json();
      if (!response.ok) {
        setRepositories([]);
        setPageCount(0);
        return navigate("/does/not/exist");
      }
      const newPageCount = PaginationService.getGitHubPageCount(response);
      const commitsInfo = await PaginationService.getCommitsInfo(response);
      setRepositories(commitsInfo);
      setCommitsInfoLength(commitsInfo.length);
      if (newPageCount !== -1) {
        setPageCount(newPageCount);
      }
    })();
  }, [appContext.perPage, navigate, org, pageOffset, repo]);

  const handlePageChange = (event: { selected: number }) => {
    setPageOffset(event.selected);
  };
  // const getApiError=()=>{
  //   return apiError && (
  //     <div className="alert alert-danger" role="alert">
  //       {apiError}
  //     </div>
  //   )
  // }

  const className = commitsInfoLength > 0 ? "mainDiv" : "hidden";
  return (
    <div className={className}>
      <div className="backWrapper">
        <Link to="/settings" className="settingsLink">
          <button type="button" className="settingsButton">
            Back to Settings
          </button>
        </Link>{" "}
      </div>
      <h3 className="repo-title">{props.orgName} GitHub repositories</h3>
      <h2 className="repo-title">
        {org} {repo}
      </h2>
      <div className="paginationWrapper">
        <Pagination
          pageCount={pageCount}
          handlePageChange={handlePageChange}
          pageOffset={pageOffset}
        />
      </div>
      {/* {getApiError()} */}
      <Commits className="listing" repositories={repositories} />
    </div>
  );
};
