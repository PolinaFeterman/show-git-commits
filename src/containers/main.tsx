import React, { useContext, useEffect, useState } from 'react';
import  '../css/paginate.css'
import generatedGitInfo from '../generatedGitInfo.json';
import { CommitCard }from './commitCard';

import ReactPaginate from "react-paginate";

import ReactDOM from "react-dom";


// Util for GitHub API.

function getGitHubPageCount(response:Response):number {
  const REGEX_GITHUB_HEADER_LINK = /<\S*\&page=([0-9]+)>; rel="last"/gm;
  // Extract pageCount from Link header.
  const l =response.headers.get("Link");
  if(l){
      const link = REGEX_GITHUB_HEADER_LINK.exec(l);
      if(link){
        return (Number(link[1]));
      }
  }
  return -1;
}


function Repositories({ repositories, className}: {
    repositories: Array<any>;
    className: string;
}) {
  if (repositories.length === 0) {
    return (
      <div className="alert alert-warning" role="alert">
        No results.
      </div>
    );
  }

  return (
    <div className={className}>
      {repositories.map((repository, index) => (
        <div className="repository-entry" key={index}>
          <h4>{repository.sha} {index}</h4>
        </div>
      ))}
    </div>
  );
}

export type mainProps={
    orgName:string;
    repoName:string;
    perPage: number;
}

export const Main=(props:mainProps)=> {
  const [repositories, setRepositories] = useState([]);
  const [pageOffset, setPageOffset] = useState(0);
  const [pageCount, setPageCount] = useState(1);
  const [apiError, setApiError] = useState(null);

  
    useEffect(() => {
    // Using an IIFE
    (async ()=> {
        const urlParams =new URLSearchParams({
            "per_page": props.perPage.toString(),
            "page": (pageOffset+1).toString()


          });
          const obj = {  
            method: 'GET',
            headers: {
              'Authorization': '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',
              'Content-Type': 'application/json',
              // 'Origin': '',
              // 'Host': 'api.producthunt.com'
            }
          }
        const response = await fetch(
          ` https://api.github.com/repos/${props.orgName}/${props.repoName}/commits?${urlParams}`, obj
          //`https://api.github.com/users/${props.orgName}/repos?${urlParams}`
        );
        const responseJson = await response.json();
        if (!response.ok) {
          setApiError(responseJson.message);
          setRepositories([]);
          setPageCount(0);
          return;
        }
        const newPageCount = getGitHubPageCount(response);
        setRepositories(responseJson);
        if(newPageCount != -1){
            setPageCount(newPageCount);
        }
        
    })();
  }, [pageOffset, props.perPage]);

  const handlePageChange = (event:any) => {
    console.log(event);
    // TODO Only change displayed selected page
    // when its content is loaded in useEffect.
    setPageOffset(event.selected);
  };

  return (
    <div style={{ marginTop: "1rem" }}>
      <h3 className="repo-title">{props.orgName} GitHub repositories</h3>
      <ReactPaginate
        previousLabel="Previous"
        nextLabel="Next"
        pageClassName="page-item"
        pageLinkClassName="page-link"
        previousClassName="page-item"
        previousLinkClassName="page-link"
        nextClassName="page-item"
        nextLinkClassName="page-link"
        breakLabel="..."
        breakClassName="page-item"
        breakLinkClassName="page-link"
        pageCount={pageCount}
        marginPagesDisplayed={2}
        pageRangeDisplayed={5}
        onPageChange={handlePageChange}
        containerClassName="pagination"
        activeClassName="active"
        forcePage={pageOffset}
      />
      {apiError && (
        <div className="alert alert-danger" role="alert">
          {apiError}
        </div>
      )}
      <Repositories className="listing" repositories={repositories} />
      {/* <ReactPaginate
        previousLabel="Previous"
        nextLabel="Next"
        pageClassName="page-item"
        pageLinkClassName="page-link"
        previousClassName="page-item"
        previousLinkClassName="page-link"
        nextClassName="page-item"
        nextLinkClassName="page-link"
        breakLabel="..."
        breakClassName="page-item"
        breakLinkClassName="page-link"
        pageCount={pageCount}
        marginPagesDisplayed={2}
        pageRangeDisplayed={5}
        onPageChange={handlePageChange}
        containerClassName="pagination"
        activeClassName="active"
        forcePage={pageOffset}
      /> */}
    </div>
  );
}



// export const Main=()=>{
//    const [length,setLength] = useState(0);
//     const [abs,setAbs] = useState<boolean>(true);
    
// const updateAPI = useEffect(()=>{
//    // fetch callAPI(abs);
// }, [abs]);

//     const f = () => {
//      setLength(length + 1);
//     };
//     return (
//         <>
//         <button onClick = {f}> {length}</button>
//     <div className="git-info">{
       
//         generatedGitInfo.map((info,index) => {return <div key={index}>
//             <CommitCard gitBranch={info.gitBranch} gitCommitHash={info.gitCommitHash}/></div>})}
 
//   </div>
//   </>)
//  }