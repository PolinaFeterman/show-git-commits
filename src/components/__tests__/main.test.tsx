import { BrowserRouter } from "react-router-dom";
import { AppCtx } from "../../app";
import { Main } from "../main";
import renderer from "react-test-renderer";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

beforeEach(() => {
    jest.mock('react-router', () => ({
        ...jest.requireActual('react-router'),
        useParams: jest.fn().mockReturnValue({ org: 'm3db', repo: 'm3' }),
      }))
  });
describe("Main", () => {
  test("should render elements", () => {
    render(
      <AppCtx.Provider
        value={{
          perPage: 10,
        }}
      >
        <BrowserRouter>
          <Main />
        </BrowserRouter>
      </AppCtx.Provider>
    );
    expect(screen.getByText("Back to Settings")).toBeInTheDocument();
  });
  test("should render correctly", () => {
    const tree = renderer
      .create(
        <AppCtx.Provider
          value={{
            perPage: 10,
          }}
        >
          <BrowserRouter>
            <Main />
          </BrowserRouter>
        </AppCtx.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
