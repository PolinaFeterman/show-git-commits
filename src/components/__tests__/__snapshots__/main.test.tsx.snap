// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render correctly 1`] = `
<div
  className="hidden"
>
  <div
    className="backWrapper"
  >
    <a
      className="settingsLink"
      href="/settings"
      onClick={[Function]}
    >
      <button
        className="settingsButton"
        type="button"
      >
        Back to Settings
      </button>
    </a>
     
  </div>
  <h2
    className="repo-title"
  >
    Commits from 
    , repository 
  </h2>
  <div
    className="paginationWrapper"
  >
    <ul
      className="pagination"
    >
      <li
        className="page-item disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          className="page-link "
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="prev"
          role="button"
          tabIndex="-1"
        >
          Previous
        </a>
      </li>
      <li
        className="page-item active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          className="page-link"
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="canonical"
          role="button"
          tabIndex="-1"
        >
          1
        </a>
      </li>
      <li
        className="page-item disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Next page"
          className="page-link "
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="next"
          role="button"
          tabIndex="-1"
        >
          Next
        </a>
      </li>
    </ul>
  </div>
  <div
    className="listing"
  >
    <div
      className="repository-entry"
    >
      <div
        className="commit-card"
      >
        <p>
          <strong>
            Commit author: 
          </strong>
           
        </p>
        <p>
          <strong>
            Commit date: 
          </strong>
           
          Invalid Date
        </p>
        <p>
          <strong>
            Commit message: 
          </strong>
           
        </p>
        <p>
          <strong>
            Commit url: 
          </strong>
           
          <a
            href=""
          />
        </p>
      </div>
    </div>
  </div>
</div>
`;
